sync/src/main/java/tech/pegasys/artemis/sync/FutureBlocks.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.sync;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+
+/** Holds blocks with slots that are in the future relative to our node's current slot */
+class FutureBlocks {
+  private final NavigableMap<UnsignedLong, Set<BeaconBlock>> queuedFutureBlocks =
+      new ConcurrentSkipListMap<>();
+
+  /**
+   * Add a block to the future blocks set
+   *
+   * @param block The block to add
+   */
+  public void add(final BeaconBlock block) {
+    queuedFutureBlocks
+        .computeIfAbsent(
+            block.getSlot(), key -> Collections.newSetFromMap(new ConcurrentHashMap<>()))
+        .add(block);
```

Comment:
We should make sure we have a follow up ticket to limit how big this collection can get.

Reply:
I sort of doubt it will be a problem.  We're only keeping 2 slots buffered right now.  But I can limit the set size just in case.

Reply:
yeah if we've validated the block signer already this will work out fine. As long as people can throw random blocks at us, all claiming to be from the same slot, as a DOS attack.